### ISSUE TITLE
Unload hook

### ISSUE THREAD

**Issue by nyarly** [direnv/direnv#129]
When leaving an .envrc directory, it'd be nice to be able to run a script to undo changes that happened when entering the dir.

For instance, I'm using chruby for use_ruby - so long as I switch from ruby project to ruby project, it's fine, but _leaving_ direnv dirs, I'm left with "whichever ruby I had last" - being able to `chruby 2.0` whenever direnv unloads would be handy.

**Comment by sknop** [comment]
Resetting the shell title is exactly the use I am looking for.

**Comment by zimbatm** [comment]
Hey @sknop, I think you can achieve what you want by changing the PS1 no ?

**Comment by zimbatm** [comment]
not tested but something like that:
```sh
in_direnv() {
  if [ -n "$DIRENV_DIR" ]; then
    echo -n -e "\033];In DirEnv\007"
  else
    echo -n -e "\033];Out of DirEnv\007"
  fi
}
PS1='\$(in_direnv) >'
```

**Comment by sknop** [comment]
Ah, did not think of using the DIRENV_DIR variable, just discovered that. Not a bad idea, I'll try that.

Thanks

**Comment by zimbatm** [comment]
Just fixed the script, it's working now

**Comment by mattford63** [comment]
I have many dirs each has a code base for a different customer. I can launch the appropriate VPN/sshuttle session on entering the dir.  But I can't unload it again :-(

**Comment by zimbatm** [comment]
Hi @mattford63, what would happen if you had two terminals that `cd` into the same customer folder?

My main issue with implementing the unload hook is that it supposes some sort of system-wide state. The second issue is, what would happen if the terminal is closed? I think this can be fixed but it would be fairly easy to leave the system in an inconsistent state.

**Comment by ChallahuAkbar** [comment]
I've got a case where this would be useful as well. In order to work on a Rails project that I'm part of, I need to run `bundle exec rake sunspot:solr:start:` when I enter the folder. However, once I leave that folder I'd like to `bundle exec rake sunspot:solr:stop`, because I don't need to leave the solr server running in the background taking up resources. Of course I could simply try to remember to stop it when I'm done with it, but that would kind of defeat the whole point of using direnv for this in the first place.

**Comment by zimbatm** [comment]
Yeah that would be useful. I haven't figured out how to deal with multiple
shells entering or leaving the folder so if you have an idea let me know.

**Comment by nyarly** [comment]
I just stumbled into another potential use case for this: I have a directory for presentation slides. In there I've got a script called "presenting.sh" that e.g. `xset s noblank` (and a bunch of other "don't blank the screen during a presentation" stuff.) 

It'd be nice to have that happen automatically and go away when I was done. Related elsewhere is the problem that closing the terminal would mean that those settings wouldn't be reverted, though.

**Comment by zimbatm** [comment]
Interesting, thanks for sharing. In this case I would say that it's the presentation software's responsibility to prevent the screen going to sleep, just like when a video player is in full screen.

**Comment by aventurella** [comment]
Ah, I didn't consider the multiple shells entering the folder problem, aka using something like tmux would present this problem. I often run my unit tests from another shell session/pane and keep my git commands in another. That would definitely be multiple shells, and I see how the issue arises now.

definitely a tricky thing to think about unless state is maintained somewhere.. blah... If the terminal crashes, the state file would effectively be inconsistent.. MORE BLAH!

Tough problem.

**Comment by zimbatm** [comment]
Most use-cases here would be solved with a process manager. direnv doesn't maintain a running process so it would have to be a new project (`dirsvc`?).

I think the realization is that we have system-level services, per-user services and now we want per-project services. In systemd, per-user services are started when the first user sessions opens and stopped when the last session closes. Similarly we want the per-project services to start when the first shell opens the project, and stop when the last shell leaves it.

Similarly to direnv it would have a shell hook. It would report directory enter and leaves to the daemon, which would then in turn start and stop the services if a specific file is found (`.svcrc`?).

```
$ dirsvc status
/home/user/src/project1 (shells: 3, services: 4)
/home/user/src/project2 (shells: 1, services: 1)
$ dirsvc stop /home/user/src/project2 # forcefully close a session if the shell has disappeared
$ cd /home/user/src/project1 # shell hook adds this shell to the session
$ dirctl status # show the running processes for this folder
* mysql: running
* redis: running
$ dirctl logs mysql
...
$ dirctl restart mysql
...
```

Potentially the hook could also be implemented in tmux which would give a better reporting facility.

The service definition format is unspecified here. It might make sense to adopt something existing like the systemd units.

A last note: the escalation between per-system, per-user and per-project is also a concern for build and runtime dependencies. This is already solved by Nix.

**Comment by nyarly** [comment]
Just chipping in: if it's a service manager, part of its job would be to
manage processes. Watching the shell that triggered it would stand firmly
in its baliwick.

**Comment by zimbatm** [comment]
Yeah makes sense, it would be easier to watch the lifetime of each shell that way. I guess tmux fits that bill already.

**Comment by skogsbrus** [comment]
I notice that this issue is very old, so maybe there is some solution to this but I haven't found it. 

I'd also like there to be an unload hook. My use case is is to run `aws-vault exec my-profile` on load, and `aws-vault clear my-profile` on unload.

**Comment by kbd** [comment]
This feature seems like it should be possible, given that direnv knows when it's unloading:

> direnv: unloading

My use case: I want to set my [kitty tab color](https://sw.kovidgoyal.net/kitty/remote-control/#kitty-set-tab-color) when I enter a direnv and reset it when I leave.

**Comment by steinybot** [comment]
Is there anyway for direnv to start a new subshell? That way we could use traps to perform cleanup.

**Comment by archcorsair** [comment]
I set some specific terminal decorations based on the directory I'm in, I'd like to be able to unload them on the way out. Would love to have an unload hook, very useful in general.

**Comment by richtong** [comment]
yes an unload hook would be great. For instance if you have Poetry for managing projects, then you want to be able to enter so .direnv has `poetry shell` in it and this works,

But when you want to cd out, you want do exit that shell. As someone said, you could do this with direnv exit where you just do a exit first, but the shell is then a different one. Hard problem surprisingly hard.

**Comment by zimbatm** [comment]
The "spawning of a subshell" approach is also a good design. It would make unloading much more straightforward since the tool doesn't have to revert the changes. It also means that aliases and functions can be supported.

I cannot convert direnv to this approach, as it would fundamentally change the tools but I can see an alternate universe where direnv would have taken that approach. Or a new tool could be written to handle this.

**Comment by tom-ricci** [comment]
I have a usecase for this: I want to start the docker daemon when I enter a project that requires docker, and stop it when I leave. Currently direnv lets me start the daemon when I enter the project, but there's no way to stop it when I leave.

Being able to run commands on unload would be great for this.

**Comment by tmeckel** [comment]
Another use case: running `pulumi login` on entering a directory. On "unload" I want to do a `pulumi logout`. Would be great to have such a feature.

**Comment by skogsbrus** [comment]
Given the age of this issue without anyone having been able to implement it, maybe it doesn't have to be a perfect solution that solves all of the problems in this thread.

> My main issue with implementing the unload hook is that it supposes some sort of system-wide state. The second issue is, what would happen if the terminal is closed? I think this can be fixed but it would be fairly easy to leave the system in an inconsistent state.

Would it be acceptable if the feature is defined as "allow users to execute a side effect on unload", full stop? i.e. direnv takes no responsibility for any states or conflicts between multiple shells - as an MVP the feature would just be "dumb" and put that burden on users with clear documentation.

**Comment by anpol** [comment]
> "allow users to execute a side effect on unload"

If you are using ZSH, try this:

1. Download script from https://gist.github.com/anpol/7a2706c2766eff454e047b7513fa511b
2. From ZSH prompt, `source ./direnv-hook-preload-postunload.zsh`
3. `cd` to a directory containing any `.envrc` file.  Observe the result:
```
direnv: postload from to /path/to/envrc/directory
direnv: postload_func to /path/to/envrc/directory
```
4. `cd` to another directory, not containing an `.envrc` file.  Observe the result:
```
direnv: preunload from /path/to/envrc/directory to
direnv: preunload_func /path/to/envrc/directory to
```

That was a demonstration that you definitely can write your own ZSH functions to handle direnv load/unload event.

-----

The documentation looks like:
1. Edit your  `.zshrc` and do the following:
3. Paste `./direnv-hook-preload-postunload.zsh` without examples.
2. Write your own function for handling an unload event.
3. Add that function to `direnv_preunload_functions` array.

**Comment by aae42** [comment]
Would like to use it to clean up containers in a shared container runtime

**Comment by anadon** [comment]
Ok.  I have read through all this and I think I have a solution drawing from some old Unix lore and newer maths.  Many programs from back in the day, and even current ones like Firefox use lockfiles to reliably control multiple processes.  In those lockfiles, we'd want to track all active instances of Direnv shells in an allowed directory.  So each access would create if need be, then obtain a lock, read the contents, append it's IP, MAC, and shell PID to the set, then release the lock.  While opening and reading, it will check all processes on the local machine for still existing; if there are processes in the lockfile but not present on the machine then this means there was an unclean exit and an "idempotent_last_exit_hook" need to be run and the non-existent processes removed from the lockfile.  Then on entering, if there are no processes a "idempotent_first_enter_hook" needs to be run.  Then the usual execution of code for entering the directory.  On exit, run a "shell_exit_hook".  Lastly, lock the lockfile, remove the shell PID from the set, and if this is the last shell exiting the directory on the machine then run the "idempotent_last_exit_hook".

Just requires file locking, and you have to be careful of simultaneous access of users on the same machine accessing the same directory and network access.  The lockfile usually has some convention like ".~direnv.lock".

It is 1AM.  I know this is dense.  Let's work through where I didn't do the best to express myself or have a malformed concept.

**Comment by bbenne10** [comment]
Thoughts: 

* `idempotent` describes the action, not the hook. I think simply `first_enter_hook` and `last_exit_hook` are fine names without needing the extra verbosity.
* I think you'll need a way to turn the layout dir into a lock dir *or* a separate function for locating where the lock goes (like `direnv_layout_dir`).
* Is the idea here to offload the killing of any long-running spawned jobs to the user? This seems like we're building a "shell-hook oriented init" and I don't know if I love that idea, but I don't know how to replicate the power with less functionality. Maybe taking inspiration from some of the minimal init systems that pre-date systemd is not a bad idea? ([This](https://github.com/bbenne10/rc), but I don't know if I think it was particularly good. I am happy to chat about it with you if you'd like) It does still require some notion of parallelization, but this could be a good starting point.

**Comment by bbenne10** [comment]
On 2/17/25 1:48 PM, Josh Marshall wrote: 
 > IMO, it is better to give them a big scary new work that makes them look at the manual than let them forget about the important implications of concurrency and robustness needed for the first and last hooks.

I dont expect new users to use this feature immediately and the complications of it all are, as you say, A LOT in the concurrent and long running cases, but I think there is a simple case that is still quite useful (ensuring files, ensuring database records exist correctly etc)

> Can you expand on what you mean in your second point? 

direnv has a rather dynamic and robust method of setting up locations for the cache already. This lock file must be equally dynamic and robust. I am unclear on how to do that.

The most immediate example is `direnv_layout_dir`, which is a function that can override the location of the direnv state directory. This directory is treated basically as opaque by direnv, so putting a file in there that is NOT generated by the layout itself seems a bad idea (I dont want things yo clash).

> This level of detail is needed due to clockless environments. 

Do you have one you are working on? Why are we optimizing for such? I think the idea of a group of services, defined as collections of functions for starting, stopping, restarting, and polling PIDs is sufficient for this task, honestly, and much simpler than a general purpose time based one. It isn't clear to me how PID tracking with timestamps will be INVOKED by the user. If it is a shell script interface ass direnvs interfaces already are), why not use that fact more deeply?

**Comment by tgrushka** [comment]
After 10+ years and a working PR (#1262), the refusal to implement unload hooks feels arbitrary. Load-time commands (e.g., `docker compose up`, `kubectl config set-context --current --namespace="$NAMESPACE"`, etc.) are allowed—why block cleanup? This is a design contradiction, not a technical limitation. A clear 'no' with rationale would be better than silent dismissal.

**Comment by bbenne10** [comment]
@tgrushka: while you are justified in your frustration with the lack of progress here, I question whether you have considered carefully how such a feature could be implemented. How do you handle concurrent access to the directory? How do you make sure the "destructor" is called only once on the last exit? #1262 is still listed as WIP and thus is not ready for merge - and likely for good reason: this is quite a difficult feature to implement with a fair amount of nuance in the finer points of it all. I'd love this feature to exist in some simple-ish way, but I am not convinced that such a simple implementation EXISTS. Doing this wrong has WORSE user experience than not implementing it and I don't think many people have considered it from that perspective.

**Comment by anadon** [comment]
I got to looking at dotenvx this week.  This might look trite, but would the approach of setting up a subshell with a passed in reserved variable and only run when that variable is detected, then on cleanup just exiting the shell?  If we use a value of the resolved absolute path to handle recursive cases, it might work.

`_direnvactivepath="$(pwd)" bash` for instance.

**Comment by bbenne10** [comment]
This proposal handles backgrounded jobs that are still owned by the controlling shell, but that doesn't seem to be enough for a general purpose hook, I think? What about teardown functions intended to be invoked at shutdown? What about disowned jobs?

A coworker of mine and I sketched out an idea by which this would work, relying on an invoked long running process that tracks other users processes in the same directory and runs some arbitrary shell script on the removal of the last one (save itself, of course). It seemed like *way* too much trouble and far too brittle for the gains.

I'm still firmly in the "if you want shutdown hooks, use manually invoked shell environments ala nix-shell and write the hooks yourself in shellHook" camp for this reason...

**Comment by zimbatm** [comment]
@bbenne10 summarized the technical situation very well.

A proper implementation needs to re-architect direnv to spawn a daemon, tmux-style, and then have every shell session open a unix socket to it. Then you can properly track open sessions. Then you need to funnel env vars back and forth, add a sub-command that lists the sessions, a way to manually kill the sessions. That's just from the top of my head. It would basically double the complexity of the project and add a lot more failure cases to deal with.

At the same time; direnv works well for me. There is too much traffic on the repo for me to deal with all the open PRs and issues, so I tend to focus on things I care about. If you want to help maintain the project with me, ping me. If it's a feature you need at work, my [focus is also available for hire](https://numtide.com/contact/).

**Comment by nyarly** [comment]
Having followed this issue for a long time, I have a couple of observations:

1) There is absolutely a use case here. There are so many contextual settings that would be great to be able to control based on automatic triggers.
2) direnv handles (very well!) a very specific subset of that use case: contextual settings: environment variables, triggers: current directory
3) Handling things like VPN connections would require something extra - you could put something into your `PS1` or equivalent, but handling it properly would really want to watch the processes that triggered the configuration, which would be a really different tool (`dirsrv`)

Based on those things, I do think there's a real call for something like a `dirsrv` tool. IMO, that's also a completely separate project. If anyone, including zimbatm, were to start it, I'd be interested, but in the meantime `direnv` is stellar, and things like `process-compose` fill a lot of gaps.

I'll also comment, zimbatm has been a great maintainer of this project, and generally good to work with as a contributor. Any of his suggestions above, I'd endorse, FWIW.

**Comment by bbenne10** [comment]
I 100% agree that there is UTILITY in the idea. Please do not take from this that I oppose a good implementation of such a feature. I am simply not sure what implementation meets the requirements without being quite complicated and maybe over generalized. Feel free to prove me wrong, though!

**Comment by deep4lpha** [comment]
Having a basic implementation would solve the issues for people immediately. Those who need a complex state management can write their own logic inside the hook to do this properly. Most of peeps need simple stuff not shutting down Cloudflare on dir exit 😛

**Comment by anadon** [comment]
Having picked up devenv recently, it might just be worth while to have a
`use devenv`.  Just need to be careful to also support devenv through
flakes.

**Comment by deep4lpha** [comment]
Thanks, I'll give it a try!

**Comment by bbenne10** [comment]
If you use nix, the right answer here to me feels like [numtide's devShell](https://github.com/numtide/devshell) (plus [nix-direnv](https://github.com/nix-community/nix-direnv) which I currently maintain). DevShell builds on similar primitives to [devenv](https://devenv.sh/), but feels more like a "normal" devshell than a from-scratch re-addressing of everything that vanilla `mkShell` solves. DevShell adds `commands` (which are a bit more generic than `mkShell` inputs), and `serviceGroups`, which compose many commands into a single executable. These do need to be manually invoked, but it feels very much more streamlined than running every dependent service by hand.  

And @deep4lpha: Feel free to work up a PR and add me as a reviewer. I'll happily review something that addresses this feature! I don't have merge bit (I take that it has been discussed, but the merging of nix-direnv and direnv hasn't happened yet), but I'd love to work through an implementation rather than discussing what *could* be done. 

As I have said: I tried. This is too complex for me to want to tackle in the general case. I am still waiting for someone to discuss anything more thought out than "Yeah! I want the ability to do stuff on exit!".

**Comment by deep4lpha** [comment]
@bbenne10 I'll give it a shot. also devenv seems to solve all my issues, but I'll check out devShell too! <3
